# League of Legends (LOL) Ranking Data

# Mount your google drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Load and Read the Dataset

Dataset: https://www.kaggle.com/datasnaek/league-of-legends

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = np.loadtxt('/content/drive/MyDrive/CSCE464/datasets/games.csv', delimiter = ",", skiprows=1)
df

array([[3.32608651e+09, 1.50427946e+12, 1.94900000e+03, ...,
        4.30000000e+01, 1.60000000e+01, 5.10000000e+01],
       [3.22956603e+09, 1.49784880e+12, 1.85100000e+03, ...,
        2.38000000e+02, 5.10000000e+01, 4.20000000e+02],
       [3.32736350e+09, 1.50436010e+12, 1.49300000e+03, ...,
        1.21000000e+02, 5.70000000e+01, 2.80000000e+01],
       ...,
       [3.32276504e+09, 1.50402986e+12, 2.15600000e+03, ...,
        5.30000000e+01, 1.10000000e+01, 1.57000000e+02],
       [3.25667537e+09, 1.49956204e+12, 1.47500000e+03, ...,
        5.10000000e+01, 9.00000000e+01, 1.14000000e+02],
       [3.31733302e+09, 1.50361275e+12, 1.44500000e+03, ...,
        1.41000000e+02, 3.10000000e+01, 1.80000000e+01]])

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/CSCE464/datasets/games.csv', sep=",")
df

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
0      3326086514  1504279457970          1949         9       1           2   
1      3229566029  1497848803862          1851         9       1           1   
2      3327363504  1504360103310          1493         9       1           2   
3      3326856598  1504348503996          1758         9       1           1   
4      3330080762  1504554410899          2094         9       1           2   
...           ...            ...           ...       ...     ...         ...   
51485  3308904636  1503076540231          1944         9       2           1   
51486  3215685759  1496957179355          3304         9       2           1   
51487  3322765040  1504029863961          2156         9       2           2   
51488  3256675373  1499562036246          1475         9       2           2   
51489  3317333020  1503612754059          1445         9       1           1   

       firstTower  firstInhibitor  firstBaron  firstDragon  ...  \
0               1               1           1            1  ...   
1               1               1           0            1  ...   
2               1               1           1            2  ...   
3               1               1           1            1  ...   
4               1               1           1            1  ...   
...           ...             ...         ...          ...  ...   
51485           2               2           0            2  ...   
51486           1               2           2            2  ...   
51487           2               2           0            1  ...   
51488           2               2           0            2  ...   
51489           1               1           1            2  ...   

       t2_towerKills  t2_inhibitorKills  t2_baronKills  t2_dragonKills  \
0                  5                  0              0               1   
1                  2                  0              0               0   
2                  2                  0              0               1   
3                  0                  0              0               0   
4                  3                  0              0               1   
...              ...                ...            ...             ...   
51485             10                  2              0               4   
51486             11                  7              4               4   
51487             10                  2              0               2   
51488             11                  3              0               1   
51489              1                  0              0               1   

       t2_riftHeraldKills  t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
0                       1      114       67       43       16       51  
1                       0       11       67      238       51      420  
2                       0      157      238      121       57       28  
3                       0      164       18      141       40       51  
4                       0       86       11      201      122       18  
...                   ...      ...      ...      ...      ...      ...  
51485                   0       55       -1       90      238      157  
51486                   1      157       55      119      154      105  
51487                   0      113      122       53       11      157  
51488                   0      154       39       51       90      114  
51489                   0       11      157      141       31       18  

[51490 rows x 61 columns]

# 2. Show all the data for the top 5 rows and bottom 5 rows

In [ ]:
df[0:5]

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
0  3326086514  1504279457970          1949         9       1           2   
1  3229566029  1497848803862          1851         9       1           1   
2  3327363504  1504360103310          1493         9       1           2   
3  3326856598  1504348503996          1758         9       1           1   
4  3330080762  1504554410899          2094         9       1           2   

   firstTower  firstInhibitor  firstBaron  firstDragon  ...  t2_towerKills  \
0           1               1           1            1  ...              5   
1           1               1           0            1  ...              2   
2           1               1           1            2  ...              2   
3           1               1           1            1  ...              0   
4           1               1           1            1  ...              3   

   t2_inhibitorKills  t2_baronKills  t2_dragonKills  t2_riftHeraldKills  \
0                  0              0               1                   1   
1                  0              0               0                   0   
2                  0              0               1                   0   
3                  0              0               0                   0   
4                  0              0               1                   0   

   t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
0      114       67       43       16       51  
1       11       67      238       51      420  
2      157      238      121       57       28  
3      164       18      141       40       51  
4       86       11      201      122       18  

[5 rows x 61 columns]

In [ ]:
df[51485:]

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
51485  3308904636  1503076540231          1944         9       2           1   
51486  3215685759  1496957179355          3304         9       2           1   
51487  3322765040  1504029863961          2156         9       2           2   
51488  3256675373  1499562036246          1475         9       2           2   
51489  3317333020  1503612754059          1445         9       1           1   

       firstTower  firstInhibitor  firstBaron  firstDragon  ...  \
51485           2               2           0            2  ...   
51486           1               2           2            2  ...   
51487           2               2           0            1  ...   
51488           2               2           0            2  ...   
51489           1               1           1            2  ...   

       t2_towerKills  t2_inhibitorKills  t2_baronKills  t2_dragonKills  \
51485             10                  2              0               4   
51486             11                  7              4               4   
51487             10                  2              0               2   
51488             11                  3              0               1   
51489              1                  0              0               1   

       t2_riftHeraldKills  t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
51485                   0       55       -1       90      238      157  
51486                   1      157       55      119      154      105  
51487                   0      113      122       53       11      157  
51488                   0      154       39       51       90      114  
51489                   0       11      157      141       31       18  

[5 rows x 61 columns]

In [ ]:
df.head() #Display top 5 data

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
0  3326086514  1504279457970          1949         9       1           2   
1  3229566029  1497848803862          1851         9       1           1   
2  3327363504  1504360103310          1493         9       1           2   
3  3326856598  1504348503996          1758         9       1           1   
4  3330080762  1504554410899          2094         9       1           2   

   firstTower  firstInhibitor  firstBaron  firstDragon  ...  t2_towerKills  \
0           1               1           1            1  ...              5   
1           1               1           0            1  ...              2   
2           1               1           1            2  ...              2   
3           1               1           1            1  ...              0   
4           1               1           1            1  ...              3   

   t2_inhibitorKills  t2_baronKills  t2_dragonKills  t2_riftHeraldKills  \
0                  0              0               1                   1   
1                  0              0               0                   0   
2                  0              0               1                   0   
3                  0              0               0                   0   
4                  0              0               1                   0   

   t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
0      114       67       43       16       51  
1       11       67      238       51      420  
2      157      238      121       57       28  
3      164       18      141       40       51  
4       86       11      201      122       18  

[5 rows x 61 columns]

In [ ]:
df.tail() #Display bottom 5 data

gameId   creationTime  gameDuration  seasonId  winner  firstBlood  \
51485  3308904636  1503076540231          1944         9       2           1   
51486  3215685759  1496957179355          3304         9       2           1   
51487  3322765040  1504029863961          2156         9       2           2   
51488  3256675373  1499562036246          1475         9       2           2   
51489  3317333020  1503612754059          1445         9       1           1   

       firstTower  firstInhibitor  firstBaron  firstDragon  ...  \
51485           2               2           0            2  ...   
51486           1               2           2            2  ...   
51487           2               2           0            1  ...   
51488           2               2           0            2  ...   
51489           1               1           1            2  ...   

       t2_towerKills  t2_inhibitorKills  t2_baronKills  t2_dragonKills  \
51485             10                  2              0               4   
51486             11                  7              4               4   
51487             10                  2              0               2   
51488             11                  3              0               1   
51489              1                  0              0               1   

       t2_riftHeraldKills  t2_ban1  t2_ban2  t2_ban3  t2_ban4  t2_ban5  
51485                   0       55       -1       90      238      157  
51486                   1      157       55      119      154      105  
51487                   0      113      122       53       11      157  
51488                   0      154       39       51       90      114  
51489                   0       11      157      141       31       18  

[5 rows x 61 columns]

# 3. Show the number of rows and columns as follows:

(51490, 61)

Rows: 51490

Columns: 61

>Hint: Use the pandas shape function: https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.shape.html

In [ ]:
df.shape

(51490, 61)

In [ ]:
print("Rows: ", df.shape[0])
print("Columns: ", df.shape[1])

Rows:  51490
Columns:  61


# 4. Show the name of the 6th column and the data type of all rows in that column

> Hint: Use the pandas columns and iloc

> https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html

> https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

In [ ]:
df.columns[5]
df.iloc[:, 5].dtype

dtype('int64')

# 5. Show the number of data and the data type of each column

> Hint: Use the pandas info() method: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51490 entries, 0 to 51489
Data columns (total 61 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   gameId              51490 non-null  int64
 1   creationTime        51490 non-null  int64
 2   gameDuration        51490 non-null  int64
 3   seasonId            51490 non-null  int64
 4   winner              51490 non-null  int64
 5   firstBlood          51490 non-null  int64
 6   firstTower          51490 non-null  int64
 7   firstInhibitor      51490 non-null  int64
 8   firstBaron          51490 non-null  int64
 9   firstDragon         51490 non-null  int64
 10  firstRiftHerald     51490 non-null  int64
 11  t1_champ1id         51490 non-null  int64
 12  t1_champ1_sum1      51490 non-null  int64
 13  t1_champ1_sum2      51490 non-null  int64
 14  t1_champ2id         51490 non-null  int64
 15  t1_champ2_sum1      51490 non-null  int64
 16  t1_champ2_sum2      51490 non-null  int6

# 6. Show / generate descriptive statistics

> Hint: Use the pandas describe() method: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html

In [ ]:
df.describe()

gameId  creationTime  gameDuration  seasonId        winner  \
count  5.149000e+04  5.149000e+04  51490.000000   51490.0  51490.000000   
mean   3.306223e+09  1.502926e+12   1832.362808       9.0      1.493552   
std    2.946096e+07  1.978026e+09    512.017696       0.0      0.499963   
min    3.214824e+09  1.496892e+12    190.000000       9.0      1.000000   
25%    3.292218e+09  1.502021e+12   1531.000000       9.0      1.000000   
50%    3.320021e+09  1.503844e+12   1833.000000       9.0      1.000000   
75%    3.327099e+09  1.504352e+12   2148.000000       9.0      2.000000   
max    3.331833e+09  1.504707e+12   4728.000000       9.0      2.000000   

         firstBlood    firstTower  firstInhibitor    firstBaron   firstDragon  \
count  51490.000000  51490.000000    51490.000000  51490.000000  51490.000000   
mean       1.471295      1.450631        1.308487      0.926510      1.442804   
std        0.520326      0.542848        0.676097      0.841424      0.569579   
min        0.000000      0.000000        0.000000      0.000000      0.000000   
25%        1.000000      1.000000        1.000000      0.000000      1.000000   
50%        1.000000      1.000000        1.000000      1.000000      1.000000   
75%        2.000000      2.000000        2.000000      2.000000      2.000000   
max        2.000000      2.000000        2.000000      2.000000      2.000000   

       ...  t2_towerKills  t2_inhibitorKills  t2_baronKills  t2_dragonKills  \
count  ...   51490.000000       51490.000000   51490.000000    51490.000000   
mean   ...       5.549466           0.985084       0.414547        1.404370   
std    ...       3.860989           1.256284       0.613768        1.224492   
min    ...       0.000000           0.000000       0.000000        0.000000   
25%    ...       2.000000           0.000000       0.000000        0.000000   
50%    ...       6.000000           0.000000       0.000000        1.000000   
75%    ...       9.000000           2.000000       1.000000        2.000000   
max    ...      11.000000          10.000000       4.000000        6.000000   

       t2_riftHeraldKills       t2_ban1       t2_ban2       t2_ban3  \
count        51490.000000  51490.000000  51490.000000  51490.000000   
mean             0.240105    108.216294    107.910216    108.690581   
std              0.427151    102.551787    102.870710    102.592145   
min              0.000000     -1.000000     -1.000000     -1.000000   
25%              0.000000     38.000000     37.000000     38.000000   
50%              0.000000     90.000000     90.000000     90.000000   
75%              0.000000    141.000000    141.000000    141.000000   
max              1.000000    516.000000    516.000000    516.000000   

            t2_ban4       t2_ban5  
count  51490.000000  51490.000000  
mean     108.626044    108.066576  
std      103.346952    102.756149  
min       -1.000000     -1.000000  
25%       38.000000     38.000000  
50%       90.000000     90.000000  
75%      141.000000    141.000000  
max      516.000000    516.000000  

[8 rows x 61 columns]

# 7. Show all the columns that contain categorical values

> Hint: Use the pandas select_dtypes() method: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.select_dtypes.html

In [ ]:
df.select_dtypes(include=int).columns
df.select_dtypes(include=object).columns

Index([], dtype='object')

# 8. Load and read the second dataset

> Chipotle dataset: https://github.com/justmarkham/DAT8/blob/master/data/chipotle.tsv

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/CSCE464/datasets/chipotle.txt', sep='\t')
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

# 9. Display the top 5 rows of data with a quantity of 3, starting from index 0.

> Hint: Use the pandas loc(), head(), and reset_index() methods

> loc(): https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html

> reset_index(): https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html

In [ ]:
df.loc[df['quantity']==3].head().reset_index()

index  order_id  quantity          item_name  \
0    409       178         3       Chicken Bowl   
1    445       193         3               Bowl   
2    689       284         3  Canned Soft Drink   
3    818       338         3      Bottled Water   
4    850       350         3  Canned Soft Drink   

                                  choice_description item_price  
0  [[Fresh Tomato Salsa (Mild), Tomatillo-Green C...    $32.94   
1  [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20   
2                                        [Diet Coke]     $3.75   
3                                                NaN     $3.27   
4                                           [Sprite]     $3.75

# 10. Remove the dollar sign in the item_price column and store it as a float data type in the new column named new_price


> Hint: Use the str and astype: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html

In [ ]:
#creates a new column using item_price, reads all values, turns into float
df['new_price'] = df['item_price'].str[1:].astype(float)
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  new_price  
0                                                   NaN     $2.39        2.39  
1                                          [Clementine]     $3.39        3.39  
2                                               [Apple]     $3.39        3.39  
3                                                   NaN     $2.39        2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
...                                                 ...        ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[4622 rows x 6 columns]

# 11. Find all data in the new_price column with values less than $5 and find their total count

> Hint: Relational Operator

In [ ]:
condition = df['new_price'] < 5
df[condition]

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
6            3         1                          Side of Chips   
...        ...       ...                                    ...   
4606      1828         1                      Canned Soft Drink   
4609      1829         1                      Canned Soft Drink   
4613      1831         1                                  Chips   
4614      1831         1                          Bottled Water   
4616      1832         1                    Chips and Guacamole   

     choice_description item_price  new_price  
0                   NaN     $2.39        2.39  
1          [Clementine]     $3.39        3.39  
2               [Apple]     $3.39        3.39  
3                   NaN     $2.39        2.39  
6                   NaN     $1.69        1.69  
...                 ...        ...        ...  
4606             [Coke]     $1.25        1.25  
4609           [Sprite]     $1.25        1.25  
4613                NaN     $2.15        2.15  
4614                NaN     $1.50        1.50  
4616                NaN     $4.45        4.45  

[1649 rows x 6 columns]

# 12. Extract data from new_price and item_name columns for Chicken Salad Bowl where value is $9 or less.

> Hint: Relational operator, logical operator, and loc

In [ ]:
condition1 = df['new_price'] <= 9
condition2 = df['item_name'] == 'Chicken Salad Bowl'
q11 = df[condition1 & condition2]
q11.shape

#or also
df.loc[(df.item_name=='Chicken Salad Bowl') & (df.new_price <= 9)]

order_id  quantity           item_name  \
44          20         1  Chicken Salad Bowl   
256        111         1  Chicken Salad Bowl   
526        220         1  Chicken Salad Bowl   
528        221         1  Chicken Salad Bowl   
529        221         1  Chicken Salad Bowl   
568        234         1  Chicken Salad Bowl   
585        240         1  Chicken Salad Bowl   
693        286         1  Chicken Salad Bowl   
702        289         1  Chicken Salad Bowl   
762        314         1  Chicken Salad Bowl   
1109       458         1  Chicken Salad Bowl   
1460       592         1  Chicken Salad Bowl   
1471       598         1  Chicken Salad Bowl   
1483       602         1  Chicken Salad Bowl   
1547       628         1  Chicken Salad Bowl   
1601       649         1  Chicken Salad Bowl   
1621       654         1  Chicken Salad Bowl   
1756       709         1  Chicken Salad Bowl   
1977       798         1  Chicken Salad Bowl   
2031       818         1  Chicken Salad Bowl   
2113       850         1  Chicken Salad Bowl   
2224       896         1  Chicken Salad Bowl   
2314       926         1  Chicken Salad Bowl   
2728      1083         1  Chicken Salad Bowl   
2858      1136         1  Chicken Salad Bowl   
2883      1145         1  Chicken Salad Bowl   
2897      1152         1  Chicken Salad Bowl   
2989      1189         1  Chicken Salad Bowl   
3033      1206         1  Chicken Salad Bowl   
3069      1222         1  Chicken Salad Bowl   
3117      1243         1  Chicken Salad Bowl   
3259      1305         1  Chicken Salad Bowl   
3261      1305         1  Chicken Salad Bowl   
3310      1327         1  Chicken Salad Bowl   
3315      1329         1  Chicken Salad Bowl   
3381      1357         1  Chicken Salad Bowl   
3411      1368         1  Chicken Salad Bowl   
3623      1450         1  Chicken Salad Bowl   
3656      1461         1  Chicken Salad Bowl   
3706      1483         1  Chicken Salad Bowl   
3712      1483         1  Chicken Salad Bowl   
3790      1517         1  Chicken Salad Bowl   
3798      1521         1  Chicken Salad Bowl   
3974      1592         1  Chicken Salad Bowl   
3997      1601         1  Chicken Salad Bowl   
4012      1606         1  Chicken Salad Bowl   
4035      1615         1  Chicken Salad Bowl   
4064      1627         1  Chicken Salad Bowl   
4205      1679         1  Chicken Salad Bowl   
4207      1679         1  Chicken Salad Bowl   
4333      1730         1  Chicken Salad Bowl   
4352      1738         1  Chicken Salad Bowl   
4439      1769         1  Chicken Salad Bowl   
4520      1797         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  new_price  
44    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  
256   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75        8.75  
526   [Roasted Chili Corn Salsa, [Black Beans, Sour ...     $8.75        8.75  
528   [Tomatillo Green Chili Salsa, [Fajita Vegetabl...     $8.75        8.75  
529   [Tomatillo Green Chili Salsa, [Fajita Vegetabl...     $8.75        8.75  
568             [Fresh Tomato Salsa, Fajita Vegetables]     $8.75        8.75  
585             [Fresh Tomato Salsa, Fajita Vegetables]     $8.75        8.75  
693   [Fresh Tomato Salsa, [Sour Cream, Cheese, Lett...     $8.75        8.75  
702   [Tomatillo Red Chili Salsa, [Black Beans, Chee...     $8.75        8.75  
762   [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
1109  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75        8.75  
1460                      [Fresh Tomato Salsa, Lettuce]     $8.75        8.75  
1471  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75        8.75  
1483  [Fresh Tomato Salsa, [Fajita Vegetables, Chees...     $8.75        8.75  
1547  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75    

# 13. Display data that includes the word "Chips" among the values stored in the "item_name" column.

> Hint: loc and str.contains() methods: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html

In [ ]:
#orders with only chips
df.loc[df.item_name == 'Chips']

#any order with chips
df.loc[df.item_name.str.contains('Chips')]

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
6            3         1                          Side of Chips   
10           5         1                    Chips and Guacamole   
14           7         1                    Chips and Guacamole   
...        ...       ...                                    ...   
4596      1826         1                    Chips and Guacamole   
4600      1827         1                    Chips and Guacamole   
4605      1828         1                    Chips and Guacamole   
4613      1831         1                                  Chips   
4616      1832         1                    Chips and Guacamole   

     choice_description item_price  new_price  
0                   NaN     $2.39        2.39  
3                   NaN     $2.39        2.39  
6                   NaN     $1.69        1.69  
10                  NaN     $4.45        4.45  
14                  NaN     $4.45        4.45  
...                 ...        ...        ...  
4596                NaN     $4.45        4.45  
4600                NaN     $4.45        4.45  
4605                NaN     $4.45        4.45  
4613                NaN     $2.15        2.15  
4616                NaN     $4.45        4.45  

[1084 rows x 6 columns]

# 14. Show only data in odd-numbered columns. For example, order_id, item_name, and item_price columns

>Hint: iloc

In [ ]:
df.iloc[:,::2]

order_id                              item_name item_price
0            1           Chips and Fresh Tomato Salsa     $2.39 
1            1                                   Izze     $3.39 
2            1                       Nantucket Nectar     $3.39 
3            1  Chips and Tomatillo-Green Chili Salsa     $2.39 
4            2                           Chicken Bowl    $16.98 
...        ...                                    ...        ...
4617      1833                          Steak Burrito    $11.75 
4618      1833                          Steak Burrito    $11.75 
4619      1834                     Chicken Salad Bowl    $11.25 
4620      1834                     Chicken Salad Bowl     $8.75 
4621      1834                     Chicken Salad Bowl     $8.75 

[4622 rows x 3 columns]

# 15. Sort data in the item_price column in descending order and reset the index

>Hint: sort_values() and reset_index(): https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

In [ ]:
df.sort_values(by=['item_price'], ascending=False).reset_index()

index  order_id  quantity            item_name  \
0      2624      1042         1     Steak Salad Bowl   
1      4419      1762         1     Steak Salad Bowl   
2      4036      1615         1     Steak Salad Bowl   
3      1825       738         1  Barbacoa Salad Bowl   
4      3115      1243         1  Carnitas Salad Bowl   
...     ...       ...       ...                  ...   
4617   3145      1254         1          Canned Soda   
4618    414       180         1          Canned Soda   
4619   3162      1262         1          Canned Soda   
4620    821       338         1          Canned Soda   
4621   1457       591         1          Canned Soda   

                                     choice_description item_price  new_price  
0     [Fresh Tomato Salsa, [Black Beans, Sour Cream,...     $9.39        9.39  
1     [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $9.39        9.39  
2     [Fresh Tomato Salsa, [Fajita Vegetables, Chees...     $9.39        9.39  
3     [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...     $9.39        9.39  
4     [Tomatillo Green Chili Salsa, [Rice, Pinto Bea...     $9.39        9.39  
...                                                 ...        ...        ...  
4617                                  [Diet Dr. Pepper]     $1.09        1.09  
4618                                       [Dr. Pepper]     $1.09        1.09  
4619                                        [Coca Cola]     $1.09        1.09  
4620                                        [Coca Cola]     $1.09        1.09  
4621                                           [Sprite]     $1.09        1.09  

[4622 rows x 7 columns]

# 16. After extracting the data where the value of the item_name column is either 'steak salad' or 'bowl', remove duplicate rows based on that column.

> Hint: relational operator, logical operator, loc, and drop_duplicates(): https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html

In [ ]:
df.loc[(df.item_name == 'Steak Salad') | (df.item_name == 'Bowl')].drop_duplicates('item_name')

order_id  quantity    item_name  \
445       193         3         Bowl   
664       276         1  Steak Salad   

                                    choice_description item_price  new_price  
445  [Braised Carnitas, Pinto Beans, [Sour Cream, C...    $22.20       22.20  
664  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     $8.99        8.99

# 17. Following Question 16 above, display only the last results under the same conditions

In [ ]:
df.loc[(df.item_name == 'Steak Salad') | (df.item_name == 'Bowl')].drop_duplicates('item_name', keep='last')

order_id  quantity    item_name  \
673        279         1         Bowl   
3502      1406         1  Steak Salad   

                                     choice_description item_price  new_price  
673   [Adobo-Marinated and Grilled Steak, [Sour Crea...     $7.40        7.40  
3502                        [[Lettuce, Fajita Veggies]]     $8.69        8.69

# 18. Find the average price of the new_price column and data greater than the average price.

> Hint: relational operator, mean(), and loc

In [ ]:
df.new_price.mean()

7.464335785374297

In [ ]:
df.loc[df.new_price > df.new_price.mean()]

order_id  quantity           item_name  \
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
8            4         1    Steak Soft Tacos   
9            5         1       Steak Burrito   
...        ...       ...                 ...   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  new_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98       10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75       11.75  
8     [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25        9.25  
9     [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25        9.25  
...                                                 ...        ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[2890 rows x 6 columns]

#19. Find data named Izze in the item_name column. Then change it to Fizzy Lizzy.

> Hint: loc

In [ ]:
df.loc[df.item_name == "Izze"].shape

(20, 6)

In [ ]:
df.loc[df.item_name == "Izze", 'item_name'] = "Fizzy Lizzy"
df.loc[df.item_name == "Fizzy Lizzy"].shape

(20, 6)

# 20. Find NaN in the choice_description column and change it to NoData


> Hint: isnull() and loc

In [ ]:
df.loc[df.choice_description.isnull(), 'choice_description'].shape
df.loc[df.choice_description.isnull(), 'choice_description'] = "NoData"
df

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                            Fizzy Lizzy   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  new_price  
0                                                NoData     $2.39        2.39  
1                                          [Clementine]     $3.39        3.39  
2                                               [Apple]     $3.39        3.39  
3                                                NoData     $2.39        2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98       16.98  
...                                                 ...        ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75        8.75  

[4622 rows x 6 columns]